File: front/assets/js/billing/types/spendings/budget.ts
export class Budget {
  email: string;
  limit: string;
  defaultEmail: string;

  static fromJSON(json: any): Budget {
    const b = new Budget();

    b.email = json.email as string;
    b.limit = json.limit as string;
    b.defaultEmail = json.default_email as string;
    return b;
  }

  static Zero(): Budget {
    const b = new Budget();
    b.email = ``;
    b.limit = `$0.00`;
    b.defaultEmail = ``;
    return b;
  }

  hasLimit(): boolean {
    return ![``, `$0.00`].includes(this.limit);
  }
}

File: front/lib/front/models/billing/budget.ex
defmodule Front.Models.Billing.Budget do
  alias InternalApi.Billing.Budget, as: GrpcBudget
  alias __MODULE__

  defstruct [:limit, :email, :default_email]

  @type t :: %Budget{
          limit: String.t(),
          email: String.t(),
          default_email: String.t()
        }

  @spec new(Enum.t()) :: t()
  def new(params \\ %{}), do: struct(Budget, params)

  @spec from_grpc(GrpcBudget.t()) :: t()
  def from_grpc(budget = %GrpcBudget{}) do
    new(
      limit: budget.limit,
      email: budget.email,
      default_email: ""
    )
  end
end
